### PR TITLE
[ROCKETMQ-28] Add TLS support to provide end-2-end security

### PR DESCRIPTION
**Pull request by lizhanhui**
https://issues.apache.org/jira/browse/ROCKETMQ-28

### CODE REVIEW

**Comment by Jaskey** [comment]
Hi , guys you miss the jira link

File: remoting/src/main/java/org/apache/rocketmq/remoting/netty/SslHelper.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+        File configFile = new File(NettySystemConfig.sslConfigFile);
+        boolean testMode = !(configFile.exists() && configFile.isFile() && configFile.canRead());
+        Properties properties = null;
+
+        if (!testMode) {
+            properties = new Properties();
+            InputStream inputStream = null;
+            try {
+                inputStream = new FileInputStream(configFile);
```
Comment (lizhanhui) [inline remoting/src/main/java/org/apache/rocketmq/remoting/netty/SslHelper.java]: This module needs to maintain compatibility with java 1.6, hence, no try-with-resource.

File: remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingServer.java [hunk 2]
```diff
@@ -128,6 +131,19 @@ public Thread newThread(Runnable r) {
                 }
             });
         }
+
```
Comment (lizhanhui) [inline remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingServer.java]: I am implementing a mechanism similar to [StartTLS](https://en.wikipedia.org/wiki/Opportunistic_TLS), this PR for now is in-progress.

**Comment by lizhanhui** [comment]
@Jaskey @dongeforever @vongosling @shroman Please review this PR.

File: remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingAbstract.java [hunk 3]
```diff
@@ -88,6 +89,11 @@
     protected Pair<NettyRequestProcessor, ExecutorService> defaultRequestProcessor;
 
     /**
+     * SSL Context.
```
Comment (vongosling) [inline remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingAbstract.java]: No need to use the same text, we'd better to remove this javadoc if we could not further describe this class :-)

File: remoting/src/main/java/org/apache/rocketmq/remoting/netty/SslHelper.java (hunk 1 above)
Comment (vongosling) [inline remoting/src/main/java/org/apache/rocketmq/remoting/netty/SslHelper.java]: @lizhanhui 
@Jaskey  agreed, client must be compatible with JDK 1.6 considering RocketMQ's massive user:-)

File: remoting/src/main/java/org/apache/rocketmq/remoting/netty/FileRegionEncoder.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.FileRegion;
+import io.netty.handler.codec.MessageToByteEncoder;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.WritableByteChannel;
+
+public class FileRegionEncoder extends MessageToByteEncoder<FileRegion> {
```
Comment (lizhanhui) [inline remoting/src/main/java/org/apache/rocketmq/remoting/netty/FileRegionEncoder.java]: I'll add some javadoc to this class.

File: remoting/pom.xml [hunk 5]
```diff
@@ -45,5 +45,22 @@
             <groupId>org.slf4j</groupId>
             <artifactId>slf4j-api</artifactId>
         </dependency>
+
+        <dependency>
+            <groupId>io.netty</groupId>
+            <artifactId>netty-tcnative-boringssl-static</artifactId>
```
Comment (zhouxinyu) [inline remoting/pom.xml]: `boringssl` is too heavy for `client` module, how about excluding it and use a more lightweight way in `client` module?

File: remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingServer.java [hunk 6]
```diff
@@ -169,6 +188,7 @@ public void initChannel(SocketChannel ch) throws Exception {
                             new NettyDecoder(),
                             new IdleStateHandler(0, 0, nettyServerConfig.getServerChannelMaxIdleTimeSeconds()),
                             new NettyConnectManageHandler(),
+                            new StartTlsHandler(),
```
Comment (zhouxinyu) [inline remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingServer.java]: This `StartTls` mechanism is great, and there is another mechanism for your reference:

For SSL 3.0 or TLS 1.0, 1.1 and 1.2:

+-------+------------------+------------------+--------+------
| 0x16  | 2 bytes version  |  2 bytes length  |  0x01  |  etc.
+-------+------------------+------------------+--------+------

The first byte always is 0x16(SSL handshake), can be used to decide use SSL or not.

Comment (lizhanhui) [inline remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingServer.java]: Nice catch. We do can simplify this process by detecting handshake per https://tools.ietf.org/html/rfc5246#section-6.2.1

Comment (lizhanhui) [inline remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingServer.java]: Updated.

**Pull request closed without merging**